examples/download.rs

```diff
@@ -64,20 +94,35 @@ impl Cli {
                 .progress_chars("#>-"),
         );
 
-        let mut device: Dfu<rusb::Context> =
-            DfuLibusb::open(&context, vid, pid, intf, alt).context("could not open device")?;
         device.with_progress({
             let bar = bar.clone();
             move |count| {
                 bar.inc(count as u64);
+                if bar.position() == file_size as u64 {
+                    bar.finish();
+                }
             }
         });
 
-        device
-            .download(file, file_size)
-            .context("could not write firmware to the device")?;
+        match device.download(file, file_size) {
+            Ok(_) => (),
+            Err(Error::LibUsb(..)) if bar.is_finished() => {
+                println!("USB error after upload; Device reset itself?");
+                return Ok(());
+            }
```

Comment:
The field `eof` is set in the state machine and is based solely on whether or not the buffer passed was empty.

```
                eof: bytes.is_empty(),
```

And the code that creates the Step::Break or Step::UsbReset only happen if and only if eof is true.

```
        if self.eof {
            log::trace!("Download loop ended");

            let descriptor = self.dfu.io.functional_descriptor();
            // If the device won't detach itself, it expects to be reset by the host as there is
            // nothing more that can be done. Otherwise it is expected to detach by itself
            log::trace!("Device will detach? {}", descriptor.will_detach);
            return if !descriptor.manifestation_tolerant && !descriptor.will_detach {
                Step::UsbReset
            } else {
                Step::Break
            };
        }
```